File: Java/src/Array_Hashing/GroupAnagrams.java
package Array_Hashing;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public class GroupAnagrams {
    public static List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> dict = new HashMap<>();

        for (String word : strs) {
            int[] count = new int[26];

            for (char ch : word.toCharArray()) {
                count[ch - 'a']++;
            }

            String signature = Arrays.toString(count);

            if (!dict.containsKey(signature)) {
                dict.put(signature, new ArrayList<>());
            }

            dict.get(signature).add(word);
        }

        return new ArrayList<>(dict.values());
    }
}

File: C#/ArraysAndHashing/GroupAnagrams.cs
using System.Collections.Generic;
using System.Linq;

public class GroupAnagramsSolution
{
    public List<List<string>> GroupAnagrams(string[] strs)
    {
        var dict = new Dictionary<string, List<string>>();

        foreach (var word in strs)
        {
            var count = new int[26];

            foreach (var ch in word)
            {
                count[ch - 'a']++;
            }

            var signature = string.Join("#", count);

            if (!dict.ContainsKey(signature))
            {
                dict[signature] = new List<string>();
            }

            dict[signature].Add(word);
        }

        return dict.Values.ToList();
    }
}
